### PR TITLE
[Djutifai] - Add enemy

### LINKED ISSUE #4: Add enemies
Enemy must:
- Walk towards player
- Damage player when almost in the player (must not go into the player). 
- Proper spawn of multiple enemies
- Die.
- Drop something (in future)
- Give exp to the player (when exp will be enabled, i guess..?)

### PR DESCRIPTION
**Pull request by Djutifai**
[Resolves #4]

What is done:
- [x] Add enemy struct
- [x] Add enemy movement
- [x] Add enemy spawner
- [x] Add enemy configuration through json config
- [x] Add enemy damage player
- [x]  Add player collision check (change to real collision when will be added)

What is not done yet:

- [ ] Enemy drop (Need to be added after <Item> issue
- [ ] Enemy collision between each-other (to prevent stacking up in one tile)

### CODE REVIEW

File: src/main.rs [hunk 1]
```diff
@@ -11,12 +11,17 @@ mod player;
 mod steering;
 mod tilemap;
 
+mod enemy;
+mod stats;
+
+#[cfg(debug_assertions)]
```
Comment (lunacys) [inline src/main.rs:17]: It won't build in release.

File: src/steering.rs [hunk 2]
```diff
@@ -152,9 +153,9 @@ impl Default for SteeringHost {
             cur_velocity: Vec2::ZERO,
             steering: Vec2::ZERO,
 
-            max_velocity: 250.0,
-            max_force: 150.0,
-            mass: 4.0,
+            max_velocity: MaxVelocity(250.0),
```
Comment (lunacys) [inline src/steering.rs:156]: You cannot query components from inside another component, in my opinion it's better to leave these as `f32`'s

File: src/tilemap.rs [hunk 3]
```diff
@@ -43,6 +43,26 @@ fn spawn_chunk(commands: &mut Commands, asset_server: &AssetServer, chunk_pos: I
             });
         }
     }
+
```
Comment (lunacys) [inline src/tilemap.rs:46]: Add a comment why we need to keep these comment lines

File: src/enemy.rs [hunk 4]
```diff
@@ -0,0 +1,287 @@
...
+                    }
+                    if let Ok(p_t) = player.get_single_mut() {
+                        let is_left = thread_rng().gen_range(0, 2);
+                        let is_up = thread_rng().gen_range(0, 2);
+                        let mut enemy_batch: Vec<(EnemyBundle, SpriteBundle, SteeringBundle, Name)> = Vec::new();
+
+                        for i in 1..thread_rng().gen_range(10, 20) {
+                            let (mut m_x, mut m_y, mut m_z): (f32, f32, f32);
+                            if is_left == 1 {
+                                m_x = thread_rng().gen_range(p_t.translation.x - 700.0, p_t.translation.x - 300.0);
```
Comment (lunacys) [inline src/enemy.rs:198]: `(p_t.translation.x - 700.0, p_t.translation.x - 300.0)`
We probably need to explicitly name these magical values and move to another struct.

File: src/enemy.rs [hunk 5]
```diff
@@ -0,0 +1,287 @@
...
+    rewards: Rewards,
+    is_elite: Option<bool>,
+    timer: Timer,
+}
+
+#[derive(Resource, Debug, Clone)]
+struct EnemySpawners(HashMap<u16, Vec<EnemySpawnComponent>>);
+
+impl EnemySpawnComponent {
+    fn new(name: String, enemy: Enemy, health: Health, damage: Damage, max_velocity: MaxVelocity, max_force: MaxForce, mass: Mass, texture: Handle<Image>,is_elite: Option<bool>, rewards: Rewards) -> Self {
```
Comment (lunacys) [inline src/enemy.rs:93]: You don't need a constructor here, you can just use struct initialization syntax, like:
```rust
EnemySpawnComponent {
    name: "my name".to_string(),
    enemy: Enemy::default(),
    ...
};
```

File: src/steering.rs (hunk 2 above)
Comment (Djutifai) [reply src/steering.rs:156]: Make sense

File: src/main.rs (hunk 1 above)
Comment (Djutifai) [reply src/main.rs:17]: done

File: src/steering.rs (hunk 2 above)
Comment (Djutifai) [reply src/steering.rs:156]: done

File: src/tilemap.rs (hunk 3 above)
Comment (Djutifai) [reply src/tilemap.rs:46]: done

File: src/steering.rs (hunk 2 above)
Comment (Djutifai) [reply src/steering.rs]: done

File: src/enemy.rs (hunk 5 above)
Comment (Djutifai) [reply src/enemy.rs]: done

File: src/enemy.rs (hunk 4 above)
Comment (Djutifai) [reply src/enemy.rs]: done

File: src/enemy.rs [hunk 6]
```diff
@@ -0,0 +1,279 @@
...
+                    // spawn elite only one time on the wave
+                    if Some(true) == spawner.is_elite && current_wave.need_wave_spawn == false {
+                        continue
+                    }
+                    if let Ok(p_t) = player.get_single_mut() {
+                        let is_left = thread_rng().gen_range(0, 2);
+                        let is_up = thread_rng().gen_range(0, 2);
+                        let mut enemy_batch: Vec<(EnemyBundle, SpriteBundle, SteeringBundle, Name)> = Vec::new();
+
+                        for i in 1..thread_rng().gen_range(ENEMY_BATCH_SIZE.min, ENEMY_BATCH_SIZE.max) {
```
Comment (lunacys) [inline src/enemy.rs:187]: You can't use `f32` to specify rng range. Cast to `i32`.

Comment (Djutifai) [reply src/enemy.rs:187]: changed to generic type

**Review by Sereja313 — APPROVED** [review]

File: src/enemy.rs [hunk 7]
```diff
@@ -0,0 +1,279 @@
...
+Die. [V]
+ */
+
+// TODO: Move to config?
+const WAVE_DURATION_SEC:u64 = 30;
+const GLOBAL_TIME_TICKER_SEC:u64 = 5;
+impl Plugin for EnemyPlugin {
+    fn build(&self, app: &mut App) {
+        app
+            .insert_resource(CurrentWave{ num: 1, timer: Timer::new(Duration::from_secs(WAVE_DURATION_SEC ),Repeating), need_wave_spawn: true })
```
Comment (Sereja313) [inline src/enemy.rs:38]: it might be worth defining `Default` trait for `CurrentWave`

Comment (Djutifai) [reply src/enemy.rs:38]: Why?
Is it possible to re-declare default values for fields? 
All I can see from [doc](https://docs.rs/bevy/latest/bevy/prelude/fn.default.html), you can only use default values for types. 
In our case we need exact values at start

Maybe I'm not getting something?

**Review by lunacys — APPROVED** [review]

File: src/enemy.rs (hunk 7 above)
Comment (lunacys) [reply src/enemy.rs:38]: I think he means you should implement `Default` for `CurrentWave` so you can initialize it like this:
```rust
// snip
.insert_resource(CurrentWave::Default())
```

Comment (Sereja313) [reply src/enemy.rs:38]: yes and also use `..default()` if you need to override some field

File: src/enemy.rs [hunk 8]
```diff
@@ -0,0 +1,279 @@
...
+                                 Name::new(spawner.name.clone() + &i.to_string())));
+                            if let Some(true) = spawner.is_elite {
+                               break;
+                            }
+                        }
+                        commands.spawn_batch(enemy_batch)
+                    }
+                }
+            }
+            if current_wave.need_wave_spawn == true {
```
Comment (Sereja313) [inline src/enemy.rs:227]: :thinking:

Comment (Djutifai) [reply src/enemy.rs:227]: i can remove the `== true` if you are about it

**Pull request merged**